### ISSUE TITLE
Support `ST` in `State`

### ISSUE THREAD

**Issue by incertia** [polysemy-research/polysemy#294]
Since we can run `State` in `IO` with `IORef`s it makes sense that we should also be able to run `State` in `ST` with `STRef`s unless I am mistaken somehow?

**Comment by incertia** [comment]
I suspect that this may be difficult to do, as the type of `runST` is
```
runST :: (forall s. ST s a) -> a
```
and we are only able to stick
```
Embed (ST s)
```
into the type level list

**Comment by KingoftheHomeless** [comment]
It's possible. Just use `runST $ runM . ... $ m`

**Comment by KingoftheHomeless** [comment]
This feels like a good first issue, actually:

**We want the following interpretations in `Polysemy.State`:**
```haskell
runStateSTRef
  :: forall s st r a
   . Embed (ST st) r
  => STRef st s
  ->Sem (State s ': r) a
  -> Sem r a
```
and
```haskell
stateToST
  :: forall s st r a
   . Embed (ST st) r
  => s
  -> Sem (State s ': r) a
  -> Sem r (s, a)
```

These should be easy to implement by using `runStateIORef` and `stateToIO` as references, adapting them to instead use `ST`/`Ref`. Don't forget to adapt the documentation too! (You can skip the warnings about it not being atomic).

**Comment by incertia** [comment]
Perhaps I should've elaborated a bit more. I did try the implementation you suggested but ended up with this:

```
> m = pure () :: Sem '[State Int, Embed (ST st)] ()
> :t runST $ runM $ stateToST 0 $ m

<interactive>:1:16: error:
    • Couldn't match type ‘Polysemy.Internal.Union.IndexOf
                             '[Embed (ST s)]
                             (Polysemy.Internal.Union.Found '[Embed (ST s)] (Embed (ST st0)))’
                     with ‘Embed (ST st0)’
        arising from a use of ‘stateToST’
      The type variable ‘st0’ is ambiguous
    • In the expression: stateToST 0
      In the second argument of ‘($)’, namely ‘stateToST 0 $ m’
      In the second argument of ‘($)’, namely ‘runM $ stateToST 0 $ m’
> :t m & stateToST 0 & runM & runST

<interactive>:1:26: error:
    • Couldn't match type ‘ST st0 (Int, ())’ with ‘forall s. ST s b’
      Expected type: ST st0 (Int, ()) -> b
        Actual type: (forall s. ST s b) -> b
    • In the second argument of ‘(&)’, namely ‘runST’
      In the expression: m & stateToST 0 & runM & runST
> :i stateToST
stateToST ::
  forall s st (r :: [(* -> *) -> * -> *]) a.
  Member (Embed (ST st)) r =>
  s -> Sem (State s : r) a -> Sem r (s, a)
```

Their implementation is fairly standard:

```haskell
runStateSTRef :: forall s st r a
               . Member (Embed (ST st)) r
              => STRef st s
              -> Sem (State s ': r) a
              -> Sem r a
runStateSTRef ref = interpret $ \case
  Get   -> embed @(ST st) $ readSTRef ref
  Put s -> embed @(ST st) $ writeSTRef ref s

stateToST :: forall s st r a
           . Member (Embed (ST st)) r
          => s
          -> Sem (State s ': r) a
          -> Sem r (s, a)
stateToST s sem = do
  ref <- embed @(ST st) $ newSTRef s
  res <- runStateSTRef ref sem
  end <- embed @(ST st) $ readSTRef ref
  return (end, res)
```

Did I perhaps make a mistake somewhere?

**Comment by KingoftheHomeless** [comment]
`stateToST` can't figure out what internal state type it should use. I admit, I didn't think of that.
However, that ambiguity should get resolved by the plugin. Are you using it? (Of course, if you're hacking in `polysemy` directly you can't use the plugin). 

Without using the plugin, this should also work:
```haskell
runST (runM $ stateToST @_ @st 0 $ m :: forall st. ST st (Int, ()) )
```

**Comment by incertia** [comment]
That does indeed solve the issue! We do it as follows though:

```haskell
m :: forall st. Sem '[State Int, Embed (ST st)] ()
m = pure ()

n :: forall st. ST st (Int, ())
n = runM $ stateToST @_ @st 0 m
```

We have to explicitly bind
```haskell
runM $ stateToST @_ @st 0 $ m :: forall st. ST st (Int, ())
```
to a name otherwise the `st` type is not in scope

**Comment by incertia** [comment]
This motivates something of the form
```haskell
runMExistential :: forall m st a. Monad (m st) => Sem '[Embed (m st)] a -> m st a
```
but since this also needs to be the same `st` as the one passed to `stateToST`, it seems very difficult because `runMExistential` will not have access to `stateToST`.

**Comment by KingoftheHomeless** [comment]
I don't understand the point of `runMExistential`; it's just a constrained version of `runM`. If the monad has a parametrized type variable you need to provide, you can simply do something like `runM @(ST s)`

**Comment by KingoftheHomeless** [comment]
Addressed by #295.
